File: python/src/ttio/spectrum.py
"""``Spectrum`` — generic multi-channel spectrum base class."""
from __future__ import annotations

from dataclasses import dataclass, field

from .axis_descriptor import AxisDescriptor
from .signal_array import SignalArray


@dataclass(slots=True)
class Spectrum:
    """Generic multi-channel 1-D spectrum with per-scan metadata.

    Holds an ordered mapping of named :class:`SignalArray` objects and
    the coordinate axes that index them, plus per-scan metadata
    (index position, scan time, optional precursor info for tandem
    MS).

    Concrete subclasses (:class:`MassSpectrum`, :class:`NMRSpectrum`)
    add their own typed metadata.

    Parameters
    ----------
    signal_arrays : dict[str, SignalArray]
        Named signal arrays keyed by channel name
        (``"mz"``, ``"intensity"``, ``"chemical_shift"``, ...).
    axes : list[AxisDescriptor]
        Coordinate axes describing the signal arrays.
    index_position : int, default 0
        Position in the parent :class:`~ttio.acquisition_run.AcquisitionRun` (0-based).
    scan_time_seconds : float, default 0.0
        Scan time in seconds from run start.
    precursor_mz : float, default 0.0
        Precursor m/z for tandem MS. ``0`` if not tandem.
    precursor_charge : int, default 0
        Precursor charge state. ``0`` if unknown.

    Notes
    -----
    API status: Stable.

    HDF5 representation: each spectrum is an HDF5 group whose
    immediate children are ``SignalArray`` sub-groups plus scalar
    attributes for the metadata fields.

    Cross-language equivalents
    --------------------------
    Objective-C: ``TTIOSpectrum`` · Java:
    ``global.thalion.ttio.Spectrum``.
    """

    signal_arrays: dict[str, SignalArray] = field(default_factory=dict)
    axes: list[AxisDescriptor] = field(default_factory=list)
    index_position: int = 0
    scan_time_seconds: float = 0.0
    precursor_mz: float = 0.0
    precursor_charge: int = 0

    def signal_array(self, name: str) -> SignalArray:
        """Return the named ``SignalArray``. Raises ``KeyError`` if absent."""
        try:
            return self.signal_arrays[name]
        except KeyError as exc:
            raise KeyError(
                f"no such signal array {name!r}; have {sorted(self.signal_arrays)}"
            ) from exc

    def has_signal_array(self, name: str) -> bool:
        """Return ``True`` iff ``name`` is in ``signal_arrays``."""
        return name in self.signal_arrays

    def signal_array_names(self) -> list[str]:
        """Return the channel names in insertion order."""
        return list(self.signal_arrays.keys())

    def __len__(self) -> int:
        if not self.signal_arrays:
            return 0
        return min(len(c) for c in self.signal_arrays.values())

File: java/src/main/java/global/thalion/ttio/Spectrum.java
/*
 * TTI-O Java Implementation
 * Copyright (c) 2026 The Thalion Initiative
 * SPDX-License-Identifier: LGPL-3.0-or-later
 */
package global.thalion.ttio;

import java.util.List;
import java.util.Map;

/**
 * Base class for any spectrum. Holds an ordered map of named
 * {@link SignalArray}s plus the coordinate axes that index them,
 * the spectrum's position in its parent run, scan time, and
 * optional precursor info for tandem MS.
 *
 * <p>Concrete subclasses ({@link MassSpectrum}, {@link NMRSpectrum},
 * {@link NMR2DSpectrum}, {@link Chromatogram}) add their own typed
 * metadata.</p>
 *
 * <p><b>HDF5 representation:</b> each spectrum is an HDF5 group
 * whose immediate children are {@code SignalArray} sub-groups
 * (one per named array) plus scalar attributes for the metadata
 * fields.</p>
 *
 * <p><b>API status:</b> Stable.</p>
 *
 * <p><b>Cross-language equivalents:</b> Objective-C
 * {@code TTIOSpectrum}, Python
 * {@code ttio.spectrum.Spectrum}.</p>
 *
 *
 */
public class Spectrum {
    private final Map<String, SignalArray> signalArrays;
    private final List<AxisDescriptor> axes;
    private final int indexPosition;
    private final double scanTimeSeconds;
    private final double precursorMz;
    private final int precursorCharge;

    /**
     * Primary constructor.
     *
     * @param signalArrays  named signal arrays (copied defensively)
     * @param axes          coordinate axes (copied defensively)
     * @param indexPosition position in the parent AcquisitionRun (0-based)
     * @param scanTimeSeconds scan time in seconds from run start
     * @param precursorMz   precursor m/z for tandem MS; 0 if not tandem
     * @param precursorCharge precursor charge state; 0 if unknown
     */
    public Spectrum(Map<String, SignalArray> signalArrays,
                    List<AxisDescriptor> axes,
                    int indexPosition, double scanTimeSeconds,
                    double precursorMz, int precursorCharge) {
        this.signalArrays = signalArrays != null ? Map.copyOf(signalArrays) : Map.of();
        this.axes = axes != null ? List.copyOf(axes) : List.of();
        this.indexPosition = indexPosition;
        this.scanTimeSeconds = scanTimeSeconds;
        this.precursorMz = precursorMz;
        this.precursorCharge = precursorCharge;
    }

    /**
     * Convenience constructor for subclasses with no axes or precursor info.
     * Delegates to the 6-arg primary constructor with empty axes and zero
     * precursor fields.
     *
     * @param signalArrays  named signal arrays
     * @param indexPosition position in the parent AcquisitionRun (0-based)
     * @param scanTimeSeconds scan time in seconds from run start
     */
    public Spectrum(Map<String, SignalArray> signalArrays,
                    int indexPosition, double scanTimeSeconds) {
        this(signalArrays, List.of(), indexPosition, scanTimeSeconds, 0.0, 0);
    }

    /** @return Unmodifiable map of named signal arrays (e.g. {@code "mz"}, {@code "intensity"}). */
    public Map<String, SignalArray> signalArrays() { return signalArrays; }

    /**
     * @param name signal array name (e.g. {@code "intensity"})
     * @return     the signal array under {@code name}, or {@code null} when absent
     */
    public SignalArray signalArray(String name) { return signalArrays.get(name); }

    /** @return Unmodifiable list of axis descriptors paired with the signal arrays. */
    public List<AxisDescriptor> axes() { return axes; }

    /** @return Zero-based position of this spectrum in its parent {@link AcquisitionRun}. */
    public int indexPosition() { return indexPosition; }

    /** @return Scan time in seconds since the start of the parent run; zero for non-time-resolved spectra. */
    public double scanTimeSeconds() { return scanTimeSeconds; }

    /** @return Precursor m/z for MSn spectra; zero for MS1 / non-MS modalities. */
    public double precursorMz() { return precursorMz; }

    /** @return Precursor charge state for MSn spectra; zero when unknown or non-MS. */
    public int precursorCharge() { return precursorCharge; }
}
